fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java

```diff
@@ -182,20 +190,98 @@ private void maybeLogBuildProgress(SubspaceProvider subspaceProvider, byte[] ret
+                .setReturnedRowLimit(getLimit()); // always respectLimit in this path
+        final ScanProperties scanProperties = new ScanProperties(executeProperties.build());
+
+        return ranges.onHasNext().thenCompose(hasNext -> {
+            if (Boolean.FALSE.equals(hasNext)) {
+                return AsyncUtil.READY_FALSE; // no more missing ranges - all done
+            }
+            Range range = ranges.next();
+            Tuple rangeStart = Arrays.equals(range.begin, START_BYTES) ? null : Tuple.fromBytes(range.begin);
+            Tuple rangeEnd = Arrays.equals(range.end, END_BYTES) ? null : Tuple.fromBytes(range.end);
+
+            RecordCursor<FDBIndexedRecord<Message>> cursor =
+                    store.scanIndexRecords(srcIndex, IndexScanType.BY_VALUE, TupleRange.between(rangeStart, rangeEnd), range.begin, scanProperties);
+
+            final AtomicLong recordsScannedCounter = new AtomicLong();
+            final AtomicReference<RecordCursorResult<FDBIndexedRecord<Message>>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+            final AtomicBoolean isEmpty = new AtomicBoolean(true);
+
+            return iterateRangeOnly(store, cursor,
+                    IndexingByIndex::castCursorResultToStoreRecord,
+                    lastResult::set,
+                    isEmpty, recordsScannedCounter
+            ).thenCompose(vignore -> {
+                long recordsScannedInTransaction = recordsScannedCounter.get();
+                recordsScanned.addAndGet(recordsScannedInTransaction);
+                if (common.isTrackProgress()) {
+                    store.context.ensureActive().mutate(MutationType.ADD, scannedRecordsSubspace.getKey(),
+                            FDBRecordStore.encodeRecordCount(recordsScannedInTransaction));
+                }
+                final byte[] realEnd =
+                        isEmpty.get() ? null : lastResult.get().getContinuation().toBytes();
+
+                return rangeSet.insertRange(store.ensureContextActive(), packOrNull(rangeStart), realEnd, true)
+                        .thenApply(ignore -> !isEmpty.get());
+            });
+        });
+    }
+
+    // support rebuildIndexAsync
+    @Nonnull
+    @Override
+    CompletableFuture<Void> rebuildIndexByEntityAsync(FDBRecordStore store) {
+        AtomicReference<byte[]> nextCont = new AtomicReference<>();
+        AtomicLong recordScanned = new AtomicLong();
+        return AsyncUtil.whileTrue(() ->
+                rebuildRangeOnly(store, nextCont.get(), recordScanned).thenApply(cont -> {
+                    if (cont == null) {
+                        return false;
+                    }
+                    nextCont.set(cont);
+                    return true;
+                }), store.getExecutor());
+    }
+
+    @Nonnull
+    private CompletableFuture<byte[]> rebuildRangeOnly(@Nonnull FDBRecordStore store, byte[] cont, @Nonnull AtomicLong recordsScanned) {
```

Comment:
Was the intention to add the range information to this implementation as well? I don't see where that was added.

Reply:
It was my understanding that in the `rebuildIndex` path we build the whole index without any commit (let alone `rangeSet` subspace). I understood that FDB uses it as optimizations for very small record stores. 
Am I missing something?

Reply:
@alecgrieser This method is for building the index in one transaction, which is used by check version.

We should definitely rename this method next time. If it breaks APIs, let's break it.